Genotyper v1

yifei-men opened the pull request:
@mlin cc @orodeh  @geetduggal 

My general plan is to:

1. Pass all the curated yml test cases (after correcting for typos :/)... this is done :)
2. Profile the "improvements" in terms of loss stats / remaining losses / new problems introduced
3. Tighten the logic / code

## 2 Main Algorithm Changes:
1. allow unification lookup (in translate_genotype) for records which do not span the exact same range as the site; also relax REF lookup to any record that has overlap with site
2. logic for managing sites with multiple records that overlap with it. We handle the following 2 cases:
   1. All records are gvcf confidence records; entire site is interpreted as REF
   2. Of all overlapping records, only 1 is not a gvcf confidence record. In this case, the variant record is used as the record to be used for translate_genotype (which, by virtue of unifier v1 would handle the genotype translation elegantly)

  - The case where multiple vcf variant records span a site is still rendered as a loss call, due to inability to assert phasing information.

## Supporting changes:
- harmonized loss tracking in test/services.cc
- enforced ordering of emitted genotype calls (addresses #50)

## Currently working on:
- ~~loss tracking needs to be refactored out of `translate_genotypes` to track input records correctly (right now the tracking for gvcf confidence records is incomplete)~~ [update 12/8]
- profiling loss statistics after genotyper changes

## KIV issues:
- supporting "joining" of homozygous REF variant records (which can effectively be treated as a gvcf record)

### Update 12/8
- Added logic for checking that records span site; and gvcf ref records meet depth requirement (if any).
- Supporting test cases (join_record*.yml) added

yifei-men commented on src/genotyper.cc:
```diff
@@ -272,21 +255,19 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
     // reference allele maps if it contains the unified site
     range rng(record);
-    if (rng.contains(site.pos)) {
+    if (rng.overlaps(site.pos)) {
```
Relaxed due to assumptions made on genotyping: `record` should be the only variant record in this site, so as long as it overlaps with `site.pos`, we can assume that any other region not covered by `rng` will be just REF.

[Note: currently, cases of multiple variant records overlapping a site still goes into this code path, which I'm planning to refactor out, but those cases should be rendered as loss due to `fill_allele` routine which safeguards against multiple records]

mlin replied:
I think we can now unconditionally set `allele_mapping[0] = 0` because subsequent changes now ensure we're looking at a rep record if we get to this point, right?

yifei-men commented on src/genotyper.cc:
```diff
@@ -413,13 +391,50 @@ Status genotype_site(const genotyper_config& cfg, MetadataCache& cache, BCFData&
+                S(translate_genotypes(cfg, site, dataset, dataset_header.get(), non_gvcf_records[0].get(),
+                                  sample_mapping, genotypes, genotyped, loss_trackers));
+            } else {
+                // Multiple variant records in this range.
```
I'm planning to lift this code path off completely (ie we don't bother with `translate_genotype` when it's a case that we can't resolve). It is left here for the moment to complete loss tracking.

yifei-men commented on src/unifier.cc:
```diff
@@ -140,8 +150,6 @@ auto partition(const discovered_or_minimized_alleles& alleles) {
 // valued by frequency, and alleles not sharing at least one reference
 // position in common conflict.)
 auto prune_alleles(const minimized_alleles& alleles, minimized_alleles& pruned) {
-    assert(partition<minimized_alleles>(alleles).size() == 1);
```
Removing assert as per private discussions.

yifei-men commented on test/data/gvcf_test_cases/vcf_services.yml:
```diff
@@ -1,5 +1,5 @@
 readme: |
-  This is a gvcf testing of the unify_sites service.
+  This is a gvcf testing of the unify_sites service. It involves a case where the unified site is longer than the vcf record. This site is callable assuming that bases not in a vcf record are treated as REF.
 
```
This is a slightly strange example, since it is not gvcf, so I'm making the assumption that any regions that are not reported in the vcf will be treated as REF.

yifei-men commented:
@mlin

Most of the logic for joining records are now moved to the function `find_rep_record` ([here](https://github.com/dnanexus-rnd/GLnexus/blob/ymen-genotyper-v1/src/genotyper.cc#L362)). I've added (rather verbosely) comments on its expected behavior (for v1).

I'm returning `ans` in `find_rep_record` as a vector (although right now the ans is either 0 or 1 record) because I'm not entirely sure if it might make sense for multiple "representative records" to be relevant in future iterations. I'm working with the assumption that all records that are found by `find_rep_record` will be passed downstream into `translate_genotypes`.

mlin commented on include/types.h:
```diff
@@ -147,11 +147,38 @@ struct range {
     bool within(const range& r) const noexcept { return rid == r.rid && beg >= r.beg && end <= r.end; }
     bool contains(const range& r) const noexcept { return r.within(*this); }
 
+    bool spanned_by(std::vector<range>& record_rngs) const noexcept {
```
Not a big deal but I find the implementation of `spanned_by` is a little bit convoluted, calling a private member function that calls back out to other members...I wonder can we think of ways to simplify. (Not urgent, can file as a future clean-up)

mlin commented on src/genotyper.cc:
```diff
@@ -347,14 +335,96 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
+///      ans contains the 1 vcf variant record, failed_ref_depth updated accordingly, no change to genotypes
+/// Records span entire range, and consists of >1 vcf variant record
+///      Empty ans is returned, failed_ref_depth updated accordingly, genotypes for all samples in this dataset are updated to be lost calls with RNC=LostAllele
+
```
As I understand it `find_rep_record` is meant to "return" either zero or one records, currently in a `vector<shared_ptr<bcf1_t>> ans` of length either zero or one. You could simplify this to `shared_ptr<bcf1_t>& ans` which is set on exit to either `nullptr` or filled. (Unless, do you envision future cases where you'll want to produce more than one record?)

mlin commented on src/genotyper.cc:
```diff
@@ -272,21 +264,19 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
 
     // map the bcf1_t alt alleles according to unification
     // this placeholder algorithm can do this only if the record covers
     // exactly the same reference range as the unified site.
```
you can update this comment :)

mlin commented on src/genotyper.cc:
```diff
@@ -418,13 +488,25 @@ Status genotype_site(const genotyper_config& cfg, MetadataCache& cache, BCFData&
+
+        vector<bool> failed_ref_depth(bcf_nsamples, false);
+        vector<shared_ptr<bcf1_t>> rep_records;
+        find_rep_record(cfg, site, records, dataset, dataset_header.get(), bcf_nsamples, sample_mapping, genotypes, failed_ref_depth, rep_records);
```
Make sure to check the return Status

mlin commented on src/genotyper.cc:
```diff
@@ -418,13 +488,25 @@ Status genotype_site(const genotyper_config& cfg, MetadataCache& cache, BCFData&
+        if(genotypes[2*i] > genotypes[2*i + 1]) {
+            swap(genotypes[2*i], genotypes[2*i+1]);
+        }
+    }
```
:+1:

mlin commented on src/genotyper.cc:
```diff
@@ -347,14 +335,96 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
-Status genotype_site(const genotyper_config& cfg, MetadataCache& cache, BCFData& data,
-                     const unified_site& site,
+        for (int i=0; i<bcf_nsamples; i++) {
+            // TODO: Consider support for hom ref records
```
It will be interesting to look at the residuals for the 8-samples to see if that case arises.

mlin commented:
Is the `genotyped` vector (and the part of `translate_genotypes that checks it) now basically unnecessary?

mlin commented:
Last comment for the future: fairly soon after V1, I think we'll want to add the ability to generate per-sample allele depth information in our output VCF file. For a given allele in the unified site, this would be the minimum of the depth of the corresponding allele in the representative input record, and the depths at each reference padding position. To make the latter available I'm thinking `find_rep_record` might produce something like `vector<int> min_padding_depth` in place of `vector<bool> failed_ref_depth`. Representing this would also allow you to centralize where you ask "is there sufficient depth?" which is split in two different places right now. One has to figure out the right representation when there's no padding needed, though.

Anyway, this is mainly to keep in mind for the next steps.

yifei-men commented:
Addressed the inline PR comments.

> Is the genotyped vector (and the part of translate_genotypes that checks it) now basically unnecessary?

It is indeed redundant now since `find_rep_record` already has the logic for marking sites with >1 vcf record as a loss. Removed the `genotyped` vector. This is based on the assumption that the same sample will not appear in multiple datasets (which I think we already guarantee).



> I'm thinking find_rep_record might produce something like vector<int> min_padding_depth in place of vector<bool> failed_ref_depth.

Implemented this, with value set as -1 for samples without reference padding, and a predicate function `sufficient_ref` in `AlleleDepthHelper` to handle interpretation of `min_allele_depth` on a per sample basis.


"Returning" `ans` in `find_rep_record` as a single `shared_ptr<bcf1_t>` instead of a vector since in v1, and in the immediate future I think we'll be restricting us to 1 representative record per site per sample.

mlin commented on src/genotyper.cc:
```diff
@@ -216,18 +220,21 @@ Status LossTracker::get(loss_stats& ans) const noexcept {
+
+            int n_calls = !bcf_gt_is_missing(gt[i*2]) + !bcf_gt_is_missing(gt[i*2 + 1]);
+            loss.add_call_for_site(rng, n_calls, is_gvcf_ref_record(cfg, record.get()));
+        }
```
Need to `free(gt)` here...hmm, I wonder why valgrind did not notice this...

mlin commented on src/genotyper.cc:
```diff
@@ -347,14 +337,105 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
+                min_ref_depth[mapped_sample] = ref_depth;
+            } else {
+                min_ref_depth[mapped_sample] = min(min_ref_depth[mapped_sample], ref_depth);
+            }
```
I think there might be a bug here in the case where we see a reference confidence record, followed by a variant record. When we get to the variant record, `ref_depth` be set to -1 by `get_gvcf_depth`, and we'll use that to overwrite an existing value of `min_ref_depth[mapped_sample]` from the previous reference confidence record. It looks like the join_records_insufficient_ref_depth.yml doesn't exercise this particular case.

mlin commented on src/genotyper.cc:
```diff
@@ -114,6 +97,27 @@ class AlleleDepthHelper {
         }
     }
+
+    // Whether there is sufficient ref coverage, expect ref coverate
```
typo

mlin commented on src/genotyper.cc:
```diff
@@ -272,21 +278,18 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
-                if (p != site.unification.end()) {
-                    allele_mapping[i] = p->second;
-                }
+    // checking for vlaid dna regex match
```
typo

mlin commented:
The `genotyped` vector is still declared, can delete that.

The comment at the very top of genotyper.cc (discussing the placeholder algorithm) can be updated/removed.

Please squash and merge after addressing the above! Great work!

yifei-men commented:
Addressed PR comments above.

Squash merged as ab687ef5e0b612d81c317d62280cfb5794c6179f
